core/src/main/java/io/grpc/internal/TransportSet.java

```diff
@@ -228,36 +246,42 @@ public void run() {
   }
 
   /**
-   * Shut down all transports, may run callback inline.
+   * Shut down all transports, stop creating new streams, but existing streams will continue.
+   *
+   * <p>May run callback inline.
    */
   final void shutdown() {
-    UncancellableTransportFuture savedActiveTransportFuture;
+    ManagedClientTransport savedActiveTransport;
     boolean runCallback = false;
     synchronized (lock) {
       if (shutdown) {
         return;
       }
       shutdown = true;
-      savedActiveTransportFuture = activeTransportFuture;
-      activeTransportFuture = NULL_VALUE_FUTURE;
+      savedActiveTransport = activeTransport;
+      activeTransport = null;
       if (transports.isEmpty()) {
         runCallback = true;
+        Preconditions.checkState(reconnectTask == null, "Should have no reconnectTask scheduled");
+        Preconditions.checkState(delayedTransport == null, "Should have no delayedTransport");
+      }  // else: the callback will be run once all transports have been terminated
+      if (delayedTransport == null) {
+        // We DO NOT cancel reconnectTask if delayedTransport is there, because there may still be
+        // pending streams in it.  Eventually a real treansport will be set to delayedTransport and
+        // the pending streams can be handled by the real transport.
+        cancelReconnectTask();
       }
-      if (reconnectTask != null) {
-        reconnectTask.cancel(false);
-      }
-      // else: the callback will be run once all transports have been terminated
     }
-    if (savedActiveTransportFuture != null) {
-      if (savedActiveTransportFuture.isDone()) {
-        try {
-          // Should not throw any exception here
-          savedActiveTransportFuture.get().shutdown();
-        } catch (Exception e) {
-          throw Throwables.propagate(e);
-        }
-      } else {
-        savedActiveTransportFuture.set(null);
+    if (savedActiveTransport != null) {
+      savedActiveTransport.shutdown();
+    }
+    synchronized (lock) {
+      // If savedActiveTransport was delayedTransport, it may have stopped reconnectTask from being
```

Comment:
Note that if you kept this, you would need to do `runCallback = true` within the `if (transports.isEmpty())` block.